aws_lambda_builders/workflows/nodejs_npm_esbuild/workflow.py

```diff
@@ -94,18 +99,63 @@ def actions_with_bundler(
         :rtype: list
         :return: List of build actions to execute
         """
+        # pylint: disable=R0915
+        # pylint: disable=R0914
         lockfile_path = osutils.joinpath(source_dir, "package-lock.json")
         shrinkwrap_path = osutils.joinpath(source_dir, "npm-shrinkwrap.json")
 
-        copy_action = CopySourceAction(source_dir, scratch_dir, excludes=self.EXCLUDED_FILES)
+        excluded = self.EXCLUDED_FILES + tuple("node_modules")
+        actions = [CopySourceAction(source_dir, scratch_dir, excludes=excluded)]
+
+        subprocess_node = SubprocessNodejs(osutils, self.executable_search_paths, which=which)
+        esbuild_check_version = EsbuildCheckVersionAction(scratch_dir, subprocess_esbuild)
+        esbuild_with_deps = EsbuildBundleAction(scratch_dir, artifacts_dir, bundler_config, osutils, subprocess_esbuild)
+        esbuild_skip_deps = EsbuildBundleAction(
+            scratch_dir,
+            artifacts_dir,
+            bundler_config,
+            osutils,
+            subprocess_esbuild,
+            subprocess_node,
+            skip_deps=True,
+        )
 
         if osutils.file_exists(lockfile_path) or osutils.file_exists(shrinkwrap_path):
             install_action = NodejsNpmCIAction(scratch_dir, subprocess_npm=subprocess_npm)
         else:
             install_action = NodejsNpmInstallAction(scratch_dir, subprocess_npm=subprocess_npm, is_production=False)
 
-        esbuild_action = EsbuildBundleAction(scratch_dir, artifacts_dir, bundler_config, osutils, subprocess_esbuild)
-        return [copy_action, install_action, esbuild_action]
+        if self.download_dependencies:
```

Comment:
I re-wrote the workflow to make it sequential instead of using nested if-else branches. Let me know if this is more readable, otherwise I'll revert. I've also attached the truth table this is based on for reference.

download_dependencies | dependencies_dir | combine_dependencies | Actions | Outcome | SAM CLI Use Case
-- | -- | -- | -- | -- | --
TRUE | Specified | TRUE | NPM Install. Copy dependencies from artifact into deps folder. Bundle. | artifact dir contains dependencies and source. deps dir contains dependencies. | Initial incremental build.
TRUE | Specified | FALSE | NPM Install. Bundle only source into artifacts dir. MOVE dependencies into deps folder. | artifact dir contains source. deps dir contains dependencies. | Initial auto dependency build.
TRUE | None | TRUE | NPM Install. Bundle. (Default base case) | artifact dir contains dependencies and source. deps dir not modified. | Standard build
TRUE | None | FALSE | Not valid. Unspecified. | N/A | N/A
FALSE | Specified | TRUE | Copy dependencies from deps folder into artifact. Bundle. | artifact dir contains dependencies and source. deps dir should not be modified. | Subsequent incremental build.
FALSE | Specified | FALSE | Simply bundle only source into artifact folder. | artifact dir contains source. deps dir should not be modified. | Subsequent auto dependency build.
FALSE | None | TRUE | Not valid. Unspecified. | N/A | N/A
FALSE | None | FALSE | Not valid. Unspecified. | N/A | N/A